query-planner-js/src/buildPlan.ts

```diff
@@ -788,22 +851,129 @@ class FetchGroup {
     this._selection.forWrite().mergeIn(selection);
   }
 
-  mergeIn(toMerge: FetchGroup, mergePath: OperationPath) {
-    assert(!toMerge.isTopLevel, () => `Shouldn't merge top level group ${toMerge} into ${this}`);
-    // Note that because toMerge is not top-level, the first "level" of it's selection is going to be a typeCast into the entity type
-    // used to get to the group (because the entities() operation, which is called, returns the _Entity and _needs_ type-casting).
-    // But when we merge-in, if the point we're merging in is already the entity, then the cast can be skipped.
-    const mergePathConditionalDirectives = conditionalDirectivesInOperationPath(mergePath);
-    const selectionSet = selectionSetOfPath(mergePath, (endOfPathSet) => {
-      assert(endOfPathSet, () => `Merge path ${mergePath} ends on a non-selectable type`);
-      for (const selection of toMerge.selection.selections()) {
-        const withoutUneededFragments = removeRedundantFragments(selection, endOfPathSet.parentType, mergePathConditionalDirectives);
-        addSelectionOrSelectionSet(endOfPathSet, withoutUneededFragments);
-      }
-    });
+  canMergeChildIn(child: FetchGroup): boolean {
+    return !!child.parentRelation(this)?.path;
+  }
+
+  /**
+   * Merges a child of `this` group into it.
+   *
+   * Note that it is up to the caller to know that doing such merging is reasonable in the first place, which
+   * generally means knowing that 1) `child.inputs` are included in `this.inputs` and 2) all of `child.selection` 
+   * can safely be queried on the `this.subgraphName` subgraph.
+   *
+   * @param child - a group that must be a `child` of this, and for which the 'path in parent' (for `this`) is 
+   *   known. The `canMergeChildIn` method can be used to ensure that `child` meets those requirement.
+   */
+  mergeChildIn(child: FetchGroup) {
+    const relationToChild = child.parentRelation(this);
+    assert(relationToChild, () => `Cannot merge ${child} into ${this}: the former is not a child of the latter`);
+    const childPathInThis = relationToChild.path;
+    assert(childPathInThis, () => `Cannot merge ${child} into ${this}: the path of the former into the later is unknown`);
+    this.mergeInInternal(child, childPathInThis);
+  }
+
+  /**
+   * Merges the provided sibling (shares a common parent) of `this` group into it.
+   *
+   * @param sibling - a sibling group of `this`. Both `this` and `sibling` must share a parent but it should also be
+   * their _only_ parent. Further `this` and `sibling` must be on the same subgraph and have the same `mergeAt`.
+   */
+  mergeSiblingIn(sibling: FetchGroup) {
```

Comment:
I notice that for merging grandchildren, you've created the helper `canMergeGrandChildIn`. What are your thoughts on creating a similar helper for siblings to protect against calling this function inadvertently?

Reply:
Good point, that's probably cleaner, and it's more consistent with the other methods, so added.